### ISSUE TITLE
print statements in model block provide no output if density is nan

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/stan#1789]
Copied from @ksvanhorn's original issue for RStan in https://github.com/stan-dev/rstan/issues/274 and validated that it's a bug in Stan and is a problem for CmdStan.
#### Summary:

If an error occurs in the model block, print statements PRECEDING the error never produce any output.
#### Description:

Originally, I was trying to debug a Stan program and scattered print statements all throughout the model block, but I got no output from any of them, even though a print statement in the transformed data block did produce output. I eventually pared it down to a very small program and verified that the error occurs on a specific line. However, the print statements on the lines preceding the error never produce any output.
#### Reproducible Steps:

The following code illustrates the problem. If you comment out the assignment to z_lo, it runs and the print statements produce output. If you leave in the assignment to z_lo, you get the error 

```
Rejecting initial value:
  Gradient evaluated at the initial value is not finite.
  Stan can't start sampling from this initial value.
```

multiple times, and there is no output from the print statements.

```
model_code <- '
data {
}
transformed data {
  real lo;
  lo <- negative_infinity();
  print("Finished transforming data.");
}
parameters {
  real scale;
}
model {
  print("Starting model.");
  scale ~ normal(0, 1);
  print("A");
  {
    real z_lo;
    print("B");
    z_lo <- lo * scale;
  }
}
'

sm <- stan_model(model_code = model_code)
fit <- vb(sm, data=list())
```
#### Current Output:

```
Error in eval(expr, envir, enclos) : Rejecting initial value:
  Gradient evaluated at the initial value is not finite.
  Stan can't start sampling from this initial value.
Rejecting initial value:
  Gradient evaluated at the initial value is not finite.
  Stan can't start sampling from this initial value.
Rejecting initial value:
  Gradient evaluated at the initial value is not finite.
  Stan can't start sampling from this initial value.
Rejecting initial value:
  Gradient evaluated at the initial value is not finite.
  Stan can't start sampling from this initial value.
Rejecting initial value:
  Gradient evaluated at the initial value is not finite.
  Stan can't start sampling from this initial value.
Rejecting initial value:
  Gradient evaluated at the initial value is not finite.
  Stan can't start sampling from this initial value.
Rejecting initial value:
  Gradient evaluated at the initial value is not finite.
  Stan can't start sampling from this initial value.
Rejecting initial value:
  Gradient evaluated
```

(verified)
#### Expected Output:

Expect to see the print statements

```
Starting model.
A
B
```
#### RStan and CmdStan Version:

2.9.0
#### R Version:

3.2.1
#### Operating System:

Mac OS X El Capitan

**Comment by bob-carpenter** [comment]
Thanks @ksvanhorn for the easy-to-reproduce issue report.

I reproduced the issue in CmdStan and RStan 2.9.0.

I assigned this to @syclik because I think it's important and I think he has the best shot at tracking it down.  I'm happy to help as it involves the generated model code.  I suspect there's buffering somewhere, maybe in a stringstream, that's not getting dumped properly.  It should all be streamed out from the model, I think.
